service/kernel/data_management/train_test_split.cpp

```diff
@@ -0,0 +1,203 @@
+/** file train_test_split.cpp */
+/*******************************************************************************
+* Copyright 2020 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "data_management/data/internal/train_test_split.h"
+#include "data_management/data/numeric_table.h"
+#include "services/env_detect.h"
+#include "service_dispatch.h"
+#include "service_data_utils.h"
+#include "threading.h"
+#include "service_numeric_table.h"
+
+namespace daal
+{
+namespace data_management
+{
+namespace internal
+{
+typedef daal::data_management::NumericTable::StorageLayout NTLayout;
+
+const size_t BLOCK_CONST = 10240;
+const size_t THREADING_BORDER = 12800000;
+
+template <typename Func>
+void runBlocks(const bool inParallel, const size_t nBlocks, Func func)
+{
+    if (inParallel)
+        daal::threader_for(nBlocks, nBlocks, [&](size_t i) { func(i); });
+    else
+        for (size_t i = 0; i < nBlocks; ++i) func(i);
+}
+
+template <typename DataType, typename IdxType>
+services::Status splitColumn(NumericTable & inputTable, NumericTable & trainTable, NumericTable & testTable,
+                             const IdxType * trainIdx, const IdxType * testIdx,
+                             const size_t nTrainRows, const size_t nTestRows, const size_t iCol)
+{
+    services::Status s;
+    BlockDescriptor<DataType> origBlock, trainBlock, testBlock;
+    inputTable.getBlockOfColumnValues(iCol, 0, nTrainRows + nTestRows, readOnly, origBlock);
+    trainTable.getBlockOfColumnValues(iCol, 0, nTrainRows, readWrite, trainBlock);
+    testTable.getBlockOfColumnValues(iCol, 0, nTestRows, readWrite, testBlock);
+    const DataType * origDataPtr = origBlock.getBlockPtr();
+    DataType * trainDataPtr = trainBlock.getBlockPtr();
+    DataType * testDataPtr = testBlock.getBlockPtr();
+    DAAL_CHECK_MALLOC(origDataPtr);
+    DAAL_CHECK_MALLOC(trainDataPtr);
+    DAAL_CHECK_MALLOC(testDataPtr);
+
+    const size_t BLOCK_SIZE = BLOCK_CONST;
```

Comment:
It can be simpler, don't reinvent the wheel :)
```cpp
const size_t nTrainBlocks = nTrainRows / BLOCK_SIZE + !!(nTrainRows % BLOCK_SIZE);
...

parallel_for(size_t iblock) {
    size_t start = iblock * BLOCK_SIZE;
    size_t end = min(start + BLOCK_SIZE, nTrainRows);
    ...
}
```

Reply:
Block borders and consts computation is corrected.